confidant/authnz/userauth.py

```diff
@@ -0,0 +1,661 @@
+                # TODO: find a way to save the angular args?
+                # authomatic adds url params google auth has stripped the
+                # angular args anyway, so let's just redirect back to the
+                # index.
+                return self.redirect_to_index()
+
+        # Authomatic will have put a redirect in our response here.
+        return response
+
+
+class SamlAuthenticator(AbstractUserAuthenticator):
+    """
+    User authenticator class implementing SAML.
+    """
+
+    @property
+    def auth_type(self):
+        return 'saml'
+
+    def __init__(self):
+        self.saml_config = self._render_saml_settings_dict()
+
+    def _load_x509_for_saml(self, path):
+        """Load an X.509 certificate from a PEM file."""
+        return cryptolib.load_x509_certificate_pem_as_bare_base64(path)
+
+    def _load_rsa_for_saml(self, path, password=None):
+        """Load an RSA private key file."""
+        return cryptolib.load_private_key_pem_as_bare_base64(path,
+                                                             password=password)
+
+    def _render_saml_settings_dict(self):
+        """
+        Given the configuration present in app.config, render a settings dict
+        suitable for passing to OneLogin_Saml2_Auth() in initialization.
+        """
+
+        debug = app.config['SAML_DEBUG']
+        if debug is None:
+            debug = app.debug
+
+        root_url = app.config['SAML_CONFIDANT_URL_ROOT']
+        if not root_url:
+            raise ValueError("Must provide SAML_CONFIDANT_URL_ROOT")
+        root_url = root_url.rstrip('/')
+
+        # TODO: also support unspecified?
+        name_id_fmt = 'urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress'
+
+        # Service Provider section
+        sp_data = {
+            'entityId': root_url + '/v1/saml/metadata',
+            'assertionConsumerService': {
+                'url': root_url + '/v1/saml/consume',
+                'binding': 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST'
+            },
+            'singleLogoutService': {
+                'url': root_url + '/v1/saml/logout',
+                'binding': 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-REDIRECT'
+            },
+            'NameIDFormat': name_id_fmt,
+        }
+
+        sp_has_key = False
+        if app.config['SAML_SP_KEY_FILE']:
+            sp_has_key = True
+            sp_data['privateKey'] = self._load_rsa_for_saml(
+                app.config['SAML_SP_KEY_FILE'],
+                password=app.config.get('SAML_SP_KEY_FILE_PASSWORD'))
+        if app.config['SAML_SP_KEY']:
+            sp_has_key = True
+            sp_data['privateKey'] = app.config['SAML_SP_KEY']
+
+        if app.config['SAML_SP_CERT_FILE']:
+            sp_data['x509cert'] = self._load_x509_for_saml(
+                app.config['SAML_SP_CERT_FILE'])
+        if app.config['SAML_SP_CERT']:
+            sp_data['x509cert'] = app.config['SAML_SP_CERT']
```

Comment:
It's a little tricky because SAML, being endlessly configurable, doesn't _require_ you to sign anything. And python-saml doesn't validate anything at initialization, only raising exceptions at runtime.

If you set the `authnRequestsSigned` setting on python-saml and you then don't provide a key/cert, python-saml will raise a runtime exception when it tries to generate the authN request. 

The approach I took here was to set `authnRequestsSigned` and related options if and only if the `SAML_SP_KEY` setting is provided to confidant. So if you provide a key, all the signing options will be set, and if you don't provide a key, no signing options will be set.

I could be convinced that Confidant's settings should more closely follow the underlying python-saml settings, but I thought it seemed bad to have multiple places where you configure the signing. (e.g. I can't think of any reason you would provide a key and _not_ want signing to be enabled.)

If you don't provide a key and don't have signing enabled, then it's typically the IDP's job to reject requests. So if you try to make a SAML AuthN request to the Google IDP, for example, it will reject unsigned requests.